device-api/src/config/find.rs

```diff
@@ -36,64 +36,49 @@ pub(crate) fn find_device_files_in(
-                .map(|(core, _)| *core)
-                .collect(),
-        );
-    }
-
-    let mut found = Vec::new();
-
+    let mut fit_device_files: HashMap<DeviceFile, CoreStatus> = HashMap::new();
     for cfg in &config.cfgs {
-        let count = cfg.count();
-        let limit = match count {
-            Count::Finite(c) => c,
-            Count::All => 255, // make the loop simple by choosing a large enough number
-        };
-        'outer: for _ in 0..limit {
-            for device in devices {
-                'inner: for dev_file in device.dev_files() {
-                    if !cfg.fit(device.arch(), dev_file) {
-                        continue 'inner;
-                    }
-
-                    let used = allocated.get_mut(&device.device_index()).unwrap();
-
-                    for core in used.iter() {
-                        if dev_file.core_range().contains(core) {
-                            continue 'inner;
-                        }
-                    }
-
-                    // this dev_file is suitable
-                    found.push(dev_file.clone());
-                    used.extend(
-                        device
-                            .cores()
-                            .iter()
-                            .filter(|idx| dev_file.core_range().contains(idx)),
-                    );
-                    continue 'outer;
+        let mut found = 0;
+        for device in devices {
+            for dev_file in device.dev_files() {
+                if !cfg.fit(device.arch(), dev_file) {
+                    continue;
                 }
+                let core_idxes: Vec<&CoreIdx> = device
+                    .cores()
+                    .iter()
+                    .filter(|c| dev_file.core_range.contains(c))
+                    .collect();
+                let status = if core_idxes
+                    .iter()
+                    .all(|c| *device.statuses.get(c).unwrap() == CoreStatus::Available)
+                {
+                    CoreStatus::Available
+                } else {
+                    CoreStatus::Unavailable
+                };
+                fit_device_files.insert(dev_file.clone(), status);
+                found += 1;
             }
-
-            // Device not found
-            match count {
-                Count::All => break 'outer,
-                Count::Finite(_) => {
+        }
+        match cfg.count() {
+            Count::Finite(n) => {
+                if n > found {
                     return Err(DeviceError::device_not_found(cfg));
                 }
             }
-        }
+            Count::All => (),
+        };
+    }
+
+    let available_device_files = fit_device_files
+        .iter()
+        .filter(|(_, status)| **status == CoreStatus::Available)
+        .map(|(dev_file, _)| dev_file.clone())
+        .collect::<Vec<DeviceFile>>();
+    if available_device_files.is_empty() {
```

Comment:
I think this check is insufficient. If the config string is `npu:0:0,npu:1:0`, there would be two matching device files and both PEs should be available to continue. If my understanding is correct *each* `cfg` should yield at least one `Available` device file and the check should be moved into the loop.

Reply:
I understand that you are concerned about the situation where only one of `npu:0:0` and `npu:0:1` is available, is that correct? I intended that if at least one of device_file that fit to config exists, `find_device_files` should return that instead of return error. I'm not sure about what is more proper behavior of `find_device_files`.

Reply:
Please correct me if I misunderstood your discussion.

When npu:0:0, npu:0:1, npu:1:0, npu:1:0 are available devices, the ideal results are `[npu:0:0, npu:0:1]` or `[npu:1:0, npu:1:1]` if we we request `warboy(1)*2`. It's because it would be able to give the next caller of `find_device_files()` a chance to use either 1 fused or 2 single PEs configurations. 

As an opposite example, let's assume that `find_device_files()` returns `[npu:0:0, npu:1:0]` or `[npu:0:1, npu:1:1]`. In that case, the next caller of `find_device_files()` won't be able to use a single fused NPU because the remaining PEs belong to the different devices.

However, I think that it may require more sophisticated device placement algorithm and takes some time. So, if you are Ok, we can postpone the implementation of this device placement algorithm to the next release.

Reply:
I agree with that. I will resolve this issue at the next release.